pkg/machines/libvirt-dbus.js

```diff
@@ -534,7 +534,16 @@ const LIBVIRT_DBUS_PROVIDER = {
 
         if (connectionName) {
             return dispatch => call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'GetDomainCapabilities', ['', '', '', '', 0], { timeout, type: 'ssssu' })
-                    .done((capsXML) => {
+                    .then((capsXML) => {
+                        const arch_args = {
+                            /* emulatorbin, arch, machine, virttype, 0 */
+                            aarch64: ['', 'aarch64', 'virt', '', 0]
+                        };
+                        const arch = getHostArch(capsXML[0]);
+                        const args = arch_args[arch] || ['', '', '', '', 0];
+                        return call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'GetDomainCapabilities', args, { timeout, type: 'ssssu' });
```

Comment:
`arch` argument for GetDomainCapabilities represents not the host `arch` but the `arch` of the `Vm` you are interested on running on the host.
At least that's what the documentation says:
```
$ virsh domcapabilities --help
...
     --arch <string>  domain architecture (/domain/os/type/@arch)
```

Here you check the host architecture.

What I think we should do in this case, it call this API before opening the dialog for the vCPUs settings, and give it as parameter the domain architecture, which we know since it's part of the `vm` object.

Reply:
Hi @KKoukiou, what a timing, I was just looking into this too :D. I agree.
Apparently we already call "getDomainCapabilities" from VmOverviewTabLibvirt (https://github.com/cockpit-project/cockpit/blob/master/pkg/machines/components/vmOverviewTabLibvirt.jsx#L88). So instead of getting maxVCPU for whole hypervisor, we could get the value of maxVCPU for each VM there.

Reply:
Hi @KKoukiou 

Will Cockpit support emulate other architecture?
for example, in x86_64 machine, you can create an AArch64 VM in it?

It seems that `virsh domcapabilities` will show host arch by default (without any args)
when I got host arch, I can base on host arch to search domain arch which will be same as host arch and different machines type to get Max vCPU count of 1 domain.
If I didn't misunderstand about this (https://libvirt.org/formatdomaincaps.html#elementsCPUAllocation)

Reply:
What I want to fix is, 
By default, Cockpit will get HypervisorMaxVCPU from `GetDomainCapabilities`
But in AArch64, it will assume that you are using arch=AArch64, and machines=integratorcp
Which `HypervisorMaxVCPU` will be only 1. [1]

But nowadays AArch64 server will have lots of CPU cores in total.
Cockpit UI will limit max vCPU of each VM because of `HypervisorMaxVCPU`

[1] `<machine maxCpus='1'>integratorcp</machine>`

Reply:
Hi @skobyda 

> So instead of getting maxVCPU for whole hypervisor, we could get the value of maxVCPU for each VM there.
I'm not sure there is a limitation of maxVCPU for whole hypervisor.
But yes, we should get the value of maxVPU for each VM.

In this case, `HypervisorMaxVCPU` will be 1, and it will limit VCPU of 1 VM

if I have anything incorrect, please correct me.
Thanks

Reply:
Hi @skobyda 

I dig into `vmOverviewTabLibvirt.jsx#L88`.
It seems Cockpit only has 2 place to call "getDomainCapabilities", 1st is get Max vCPU of hypervisor, 2nd is get bootloader from it.
It seems it still used `HypervisorMaxVCPU` to determine the maximum vCPU which one vm could have.

Link [2] will limit the number of Max CPU in WebUI [3]
So maybe `HypervisorMaxVCPU` should be renamed to "DomainMaxVCPU" or something else

[1] https://github.com/cockpit-project/cockpit/blob/master/pkg/machines/components/vcpuModal.jsx#L205-L208
[2] https://github.com/cockpit-project/cockpit/blob/master/pkg/machines/components/vcpuModal.jsx#L70-L103
[3] 
![image](https://user-images.githubusercontent.com/2821179/92738638-b2551180-f3ae-11ea-9f38-20181c9188d3.png)

Reply:
by the way, the picture is under x86_64.
My Arm server is under control by OpenStack right now.
I cannot install cockpit to test it.

Reply:
> for example, in x86_64 machine, you can create an AArch64 VM in it?

Yes, you just need problem qemu emulator. Example: https://fedoraproject.org/wiki/QA:Testcase_Virt_AArch64_on_x86

> It seems that `virsh domcapabilities` will show host arch by default (without any args)
> when I got host arch, I can base on host arch to search domain arch which will be same as host arch and different machines type to get Max vCPU count of 1 domain.
> If I didn't misunderstand about this (https://libvirt.org/formatdomaincaps.html#elementsCPUAllocation)

Reply:
Hi @skobyda 

I know qemu can emulate other arch.
But Cockpit will allow to emulate other architecture or not?
I think this will be more like "design" or "policy" issue.
if yes, it might need some refactor to support emulate other arch.
if no, maybe we can start to support non-x86 platform first and only support same arch vm?

Thanks again to discuss with me.
Thanks a lot :D

Reply:
> What I want to fix is,
> By default, Cockpit will get HypervisorMaxVCPU from `GetDomainCapabilities`
> But in AArch64, it will assume that you are using arch=AArch64, and machines=integratorcp
> Which `HypervisorMaxVCPU` will be only 1. [1]
> 
> But nowadays AArch64 server will have lots of CPU cores in total.
> Cockpit UI will limit max vCPU of each VM because of `HypervisorMaxVCPU`
> 
> [1] `<machine maxCpus='1'>integratorcp</machine>`

You are right about that, that breaks functionality for aarch64 (and others architectures too like armv6l armv7l). 

The thing we are doing the wrong approach. You are right, there is no limitation for a whole hypervisor. Yet we calculate maxVCPU for a whole hypervisor by calling GetDomainCapabilities and letting it assumes host's architecture and machine type, even if each virtual machine may be emulated on different architecture and machine type (and thus each virtual machine can have different maxVCPU)

Instead, the proper solution should be calculating maxVCPU for each virtual machine separately.

This can be done by calling GetDomainCapabilities for each virtual machine and passing VM's architecture and emulated machine as "arch" and "machines" function parameters. (You can find these values in vm object in vmOverviewTab as vm.emulatedMachine and vm.arch)

Reply:
> Hi @skobyda
> 
> I know qemu can emulate other arch.
> But Cockpit will allow to emulate other architecture or not?
> I think this will be more like "design" or "policy" issue.
> if yes, it might need some refactor to support emulate other arch.
> if no, maybe we can start to support non-x86 platform first and only support same arch vm?
> 
> Thanks again to discuss with me.
> Thanks a lot :D

Right now it's not possible to set architecture or emulated machine from cockpit, but looking at virt-manager, you can set VM's architecture and emulated machine there, so it's possible to implement it for cockpit too. 
I don't think we have any design policy which limits virtual machines in cockpit only to specific architectures. The functionality which can configure is not implemented yet.

Reply:
So to specify, I think we are open to supporting other architectures for VM's too :) (it's just not implemented yet, but if there is a big use case for that, implementing it is not a problem)

Reply:
So I think we have two approach here, or we can do both in short-term and long-term.
1. Support different arch, and only support "host arch" VM (VM arch is same as host)
2. Support different arch and vm arch don't need to be same as host.

(I guess your opinion is doing both)

And it will depend on workflow of Cockpit project.
Cockpit has fast tempo in development.
When multiple arch support is fully prepared, it will have lots of merge conflict.
maybe starting from 1. and always thinking about 2. will be good approach?

Or there is some good approach that I didn't realize.
Please let me know.

Thanks a lot.
I learned a lot.

Reply:
> So I think we have two approach here, or we can do both in short-term and long-term.
> 
> 1. Support different arch, and only support "host arch" VM (VM arch is same as host)

What do you mean support different arch and only support host arch? You mean support recognizing VM with different architectures, but support the creation of MV with only host arch thru Cockpit UI.

Reply:
> What do you mean support different arch and only support host arch? You mean support recognizing VM with different architectures, but support the creation of MV with only host arch thru Cockpit UI.

for example, we can support x86_64 host with x86_64 vm right now.
and we can start to support aarch64 host with aarch64 vm in short term.
In short term, we don't need to support x86_64 host with aarch64 vm.
It will be easier to support aarch64 host for those people who want to run cockpit on aarch64 server.

we can install cockpit in aarch64 server, but the behavior is not always correct or expected.
For example, in this issue #14210, cockpit cannot create vm with more than 1 vCPU.
So we can fix the behavior in other arch host first.

> Also I don't think there would be that many merge conflicts, since domainCapabilities is the only libvirt API which has architecture as a parameter. Everything else should work the same way, independently of architecture.

I didn't have whole view of cockpit.
If you told me like that, I think you are right.

Reply:
> > What do you mean support different arch and only support host arch? You mean support recognizing VM with different architectures, but support the creation of MV with only host arch thru Cockpit UI.
> 
> for example, we can support x86_64 host with x86_64 vm right now.
> and we can start to support aarch64 host with aarch64 vm in short term.
> In short term, we don't need to support x86_64 host with aarch64 vm.
> It will be easier to support aarch64 host for those people who want to run cockpit on aarch64 server.

So that case mostly works now. If you create a VM, libvirt automatically assumes the architecture of the host, so on aarch64 host you will have aarch64 vm and same for x86_64. Only think missing to fully support that use case is that we calculate vCPU wrong, which we can fix in this PR. So once this PR is merged, I think your
> 1. Support different arch, and only support "host arch" VM (VM arch is same as host)

will be implemented.

And then supporting the creation of VMs with different arch than host can be implemented.

Reply:
yes, this PR will be for 1.

But I'm not sure that my code pass correct parameter because I'm not good at JS and I cannot found Libvirt dbus API parameter docs. 
(by the way, I refer this parameter order to write those code [1])
That is why I mentioned this will be "prototype" code. (or maybe it is correct code to use?)

[1] https://libvirt.org/html/libvirt-libvirt-domain.html#virConnectGetDomainCapabilities

Reply:
Yes.

Codewise it looks good. I just think it's better to choose a different approach and calculate the maxVCPU for each VM.

Reply:
I think this PR is a kind of workaround.
Different approach will be better to get max vCPU for vm.

Reply:
I prefer doing it the correct way now (calculate max vCPU for each VM) instead of doing a workaround now and then rewriting it to a correct way in another PR.

Thank you for discussing this with us.

Reply:
Sure!
I'm arm/arm64 developer before
and I still maintain some arm64 server right now.
that is the reason why I will notice about this.

Feel free to close this PR.
Thanks for discussing with me!

Reply:
So do want to try to work on the other approach or should I work on it and then mention you in other PR?

Reply:
I'm not good at JS.
So feel free to do it!
And I can learn from your PR!

Thanks a lot!